src/frontend/src/lib/services/utils.services.ts

```diff
@@ -0,0 +1,58 @@
+import { trackEvent } from '$lib/services/analytics.services';
+import { nullishSignOut } from '$lib/services/auth.services';
+import { toastsError } from '$lib/stores/toasts.store';
+import type { OptionIdentity } from '$lib/types/identity';
+import { errorDetailToString } from '$lib/utils/error.utils';
+import type { Identity } from '@dfinity/agent';
+import { isNullish, nonNullish } from '@dfinity/utils';
+
+/**
+ * Higher-order function to handle common authentication, tracking, and error handling patterns
+ */
+export const wrapCallWith =
+	<T extends object, R>({
+		methodToCall,
+		toastErrorMessage,
+		identity,
+		trackEventNames = {}
+	}: {
+		methodToCall: (params: T & { identity: Identity }) => Promise<R>;
+		toastErrorMessage?: string;
+		identity: OptionIdentity;
+		trackEventNames?: { success?: string; error?: string };
+	}) =>
+	async (params: Omit<T, 'identity'>): Promise<R | undefined> => {
+		if (isNullish(identity)) {
+			await nullishSignOut();
+			return undefined;
+		}
+
+		try {
+			// We know this is safe because we're adding the identity property
+			const paramsWithIdentity = {
+				...params,
+				identity
+			} as T & { identity: Identity };
+
+			const result = await methodToCall(paramsWithIdentity);
+
+			if (nonNullish(trackEventNames.success)) {
+				trackEvent({ name: trackEventNames.success });
+			}
+
+			return result;
+		} catch (err) {
+			if (nonNullish(trackEventNames.error)) {
+				trackEvent({
+					name: trackEventNames.error,
+					metadata: { error: `${errorDetailToString(err)}` }
+				});
+			}
+
+			if (nonNullish(toastErrorMessage)) {
+				toastsError({ msg: { text: toastErrorMessage }, err });
```

Comment:
no need to throw it if we handle it with a toast error

```suggestion
				toastsError({ msg: { text: toastErrorMessage }, err });
				return 
```

Reply:
I think we should, maybe the consumer need to know if there was an error or not.

Reply:
Agree, done.